NITA/lib/jnpr/toby/docs/pause.1
.\" Manpage for Toby pause/pause_on_failure feature
.\" Contact selvat@juniper.net to correct errors or typos
.TH man 1 "12 Jun 2019"
.SH FEATURE
Pause and Pause on Failure Feature
.SH SYNOPSIS
The Toby Pause & Pause on Failure feature allows users to debug toby script at time of failure while developing toby script
.SH DESCRIPTION
This page documents the Toby Pause & Pause on failure feature. This feature allows users to pause the script whenever the failure occurs and run various device commands on a given device at the time of failure to cross check whether the device is responding in a expected way or not.  This is particularly useful because it can be difficult and time consuming for engineers to try and reproduce a given issue.
.SH DEFINING PAUSE
The first step in using the pause feature is to create a yaml file to set the breakpoints at various point of failure.

.nf
.B "sample_pause_file.yaml [START]"

#(copy and paste from as needed)
# Note that even with all of the comments in this file, it is possible to copy and paste the below content and edit as needed

testcases:
     before:
         - Testcase02
         - ANY
     after:
         - Testcase01
     fail:
         - Testcase01
         - any
keywords:
     before:
         - Keyword01
     after:
         - Keyword02
         - any
     error:
         - any
     fail:
         - Keyword01
         - any

# This generic example shows a constraint where the script has to pause
.PP
.B "PAUSE USAGE"
       toby -p topology.yaml --pause sample_pause_file.yaml test.robot

.SH DEFINING PAUSE ON FAILURE
This is similar to pause feature. But user don't want to pass any yaml file. Due to that user cannot set any conditional pause. Still it would be useful to pause at any point of failure while executing a toby script during development.
.PP
.B "PAUSE ON FAILURE USAGE"
       toby -p topology.yaml --pause_on_failure test.robot

.SH SEE ALSO
toby, t, config engine, verification engine, monitoring engine
.SH BUGS
No known bugs.
.SH AUTHOR
Selvarajan Thanabal (selvat@juniper.net)


NITA/lib/jnpr/toby/tools/listeners/pause.py
'''
      Pause listener
'''
import os
import ruamel.yaml as yaml
from sys import stdout as console
from robot.libraries.BuiltIn import BuiltIn
import signal

ROBOT_LISTENER_API_VERSION = 2

def resume_execution(signal_number, frame):
    '''
    resume_execution
    '''
    print("\nResuming execution...\n")
    t.log("Resuming execution...\n")
    del signal_number   #unused hence deleting to improve pylint score
    del frame           #unused hence deleting to improve pylint score

def pause_execution():
    '''
    pause_execution
    '''
    console.write('\nToby process [ %s ] sending signal to group PID [ %s ] ' % (str(os.getpid()), os.getpgid(os.getpid())))
    t.log('Toby process [ %s ] sending signal to parent process [ %s ]\n' % (str(os.getpid()), os.getpgid(os.getpid())))
    os.kill(os.getpgid(os.getpid()), signal.SIGCONT)
    signal.signal(signal.SIGINT, resume_execution)
    print("\nPress Ctrl+C to resume...")
    signal.pause()
#########################################################
#
#  Pause Listener
#
#########################################################

class pause(object):
    '''
      Pause listener
       This listener is attached when --pause option is enabled from toby wrapper script with required YAML input file.
       Based on input YAML file, it will Pause the excecution and wait for SIGCONT to resume execution.
    '''
    ROBOT_LISTENER_API_VERSION = 2

    def __init__(self):
        self.pause_file_content = None
        console.write('Running with Pause listener attached...\n')

    def load_pause_yaml(self):
        '''
        load_pause_yaml
        '''
        if self.pause_file_content is None:
            val = BuiltIn().get_variable_value('${pause_file}')
            if val:
                try:
                    self.pause_file_content = yaml.safe_load(open(val))
                except Exception as error:
                    raise error

    def check_for_pause(self, where, when, name):
        '''
        check_for_pause
        '''
        try:
            input_list = self.pause_file_content[where][when]
            input_list = [element.lower() for element in input_list]
            if name.lower() in input_list or 'any' in input_list:
                return True
        except Exception:
            pass
        return False

    def end_keyword(self, name, attrs):
        """
            This listener method is called when keyword ends

            :param name:
                **REQUIRED** The shortname of test suite
            :param attrs:
                **REQUIRED** The attribute dictionary containing type, kwname, librname,
                 starttime, endtime, elapsedtime and status.
            :return:
                None
        """
        self.load_pause_yaml()
        if self.check_for_pause('keywords', 'after', name) or (attrs['status'] == "FAIL" and self.check_for_pause('keywords', 'fail', name)):
            print("\nPaused after Keyword: " + name)
            t.log("Paused after Keyword: " + name)
            pause_execution()

    def start_keyword(self, name, attrs):
        """
            This listener method is called when keyword starts

            :param name:
                **REQUIRED** The shortname of test suite
            :param attrs:
                **REQUIRED** The attribute dictionary containing type, kwname, librname,
                 starttime, endtime, elapsedtime and status.
            :return:
                None
        """
        self.load_pause_yaml()
        if self.check_for_pause('keywords', 'before', name):
            print("\nPaused before Keyword: " + name)
            t.log("Paused before Keyword: " + name)
            pause_execution()
            del attrs    #unused hence deleting to improve pylint score

    def start_test(self, name, attrs):
        """
            This listener method is called when test starts

            :param name:
                **REQUIRED** The shortname of test suite
            :param attrs:
                **REQUIRED** The attribute dictionary containing id, longname, starttime,
            :return:
                None
        """
        self.load_pause_yaml()
        if self.check_for_pause('testcases', 'before', name):
            print("\nPaused before Test-case: " + name)
            t.log("Paused before Test-case: " + name)
            pause_execution()
            del attrs    #unused hence deleting to improve pylint score

    def end_test(self, name, attrs):
        """
            This listener method is called when test ends

            :param name:
                **REQUIRED** The shortname of test suite
            :param attrs:
                **REQUIRED** The attribute dictionary containing id, longname, starttime,
            :return:
                None
        """
        self.load_pause_yaml()
        if self.check_for_pause('testcases', 'after', name) or (attrs['status'] == "FAIL" and self.check_for_pause('testcases', 'fail', name)):
            print("\nPaused after Test-case: " + name)
            t.log("Paused after Test-case: " + name)
            pause_execution()
